presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PushdownUtils.java

```diff
@@ -0,0 +1,152 @@
+import com.facebook.presto.spi.plan.AggregationNode;
+import com.facebook.presto.spi.plan.TopNNode;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.google.common.collect.ImmutableList;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Optional;
+
+public class PushdownUtils
+{
+    private PushdownUtils() {}
+
+    public enum ExprType
+    {
+        GROUP_BY,
+        AGGREGATE,
+    }
+
+    /**
+     * Group by field description
+     */
+    public static class GroupByColumnNode
+            extends AggregationColumnNode
+    {
+        private final VariableReferenceExpression inputColumn;
+
+        public GroupByColumnNode(VariableReferenceExpression inputColumn, VariableReferenceExpression output)
+        {
+            super(ExprType.GROUP_BY, output);
+            this.inputColumn = inputColumn;
+        }
+
+        public VariableReferenceExpression getInputColumn()
+        {
+            return inputColumn;
+        }
+
+        @Override
+        public String toString()
+        {
+            return inputColumn.toString();
+        }
+    }
+
+    /**
+     * Agg function description.
+     */
+    public static class AggregationFunctionColumnNode
+            extends AggregationColumnNode
+    {
+        private final CallExpression callExpression;
+
+        public AggregationFunctionColumnNode(VariableReferenceExpression output, CallExpression callExpression)
+        {
+            super(ExprType.AGGREGATE, output);
+            this.callExpression = callExpression;
+        }
+
+        public CallExpression getCallExpression()
+        {
+            return callExpression;
+        }
+
+        @Override
+        public String toString()
+        {
+            return callExpression.toString();
+        }
+    }
+
+    public static void checkSupported(boolean condition, String errorMessage, Object... errorMessageArgs)
+    {
+        if (!condition) {
+            throw new PinotException(PinotErrorCode.PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), String.format(errorMessage, errorMessageArgs));
+        }
+    }
+
+    public abstract static class AggregationColumnNode
+    {
+        private final ExprType exprType;
+        private final VariableReferenceExpression outputColumn;
+
+        public AggregationColumnNode(ExprType exprType, VariableReferenceExpression outputColumn)
+        {
+            this.exprType = exprType;
+            this.outputColumn = outputColumn;
+        }
+
+        public VariableReferenceExpression getOutputColumn()
+        {
+            return outputColumn;
+        }
+
+        public ExprType getExprType()
+        {
+            return exprType;
+        }
+    }
+
+    public static Optional<List<AggregationColumnNode>> computeAggregationNodes(AggregationNode aggregationNode)
+    {
+        int groupByKeyIndex = 0;
+        boolean supported = true;
```

Comment:
We can get rid of this flag variable by simply returning Optional.empty mid for-loop. In general, we do not want to introduce state when state isn't needed.